### DISCUSSION TITLE
Importing a library for use in actions

### DISCUSSION THREAD

**Discussion by NoahTheDuke** [antlr/antlr4#4178] [Q&A] [answered]
I'm working on a Clojure codebase that's parsing Clojure code using an antlr4 grammar. The output of antlr rules is strings. I'd like to convert some of these at parse time to Clojure datatypes, symbols and keywords specifically. According to the documentation: "Actions are blocks of text written in the target language and enclosed in curly braces."

Is it possible to compile the grammar with references to outside libraries and then call functions within those libraries in my grammar's actions? I want to write `keyword: KEYWORD { Clojure.var("clojure.core", "keyword").invoke($keyword.text); };` but I need [clojure.java.api](http://clojure.github.io/clojure/javadoc/clojure/java/api/package-summary.html) to be available on the classpath and imported when executing this.

Might be related to my other question in #4177, now that I've written it out.

**Comment by ericvergnaud** [comment, marked as answer]
You can use lexer:header and parser:header sections to achieve that

**Reply by NoahTheDuke** [reply]
And I'd just add the Clojure jar to the class path when I call `java -jar antlr... Clojure.g4`?

**Reply by ericvergnaud** [reply]
No, your cmd line will generate a parser, not run it. You need to integrate the generated parser in your program for it to run.

**Reply by NoahTheDuke** [reply]
Cool, thank you.
